### ISSUE TITLE
cvmfs client getting mixed up with simultaneous external server requests

### ISSUE THREAD

**Issue by DrDaveD** [cvmfs/cvmfs#3409]
Most of the external servers that OSG uses are xrootd servers with http interfaces, but at Fermilab we override the usual client configuration for osgstorage.org repositories originating at Fermilab to read directly from the dCache origin server http/webdav interface.  While investigating another issue I noticed many unexpected `switching host` messages in `/var/log/messages` when I configured a cvmfs client to read from dCache.  I reproduced this with cvmfs-client-2.11.0 on an el7 test VM at CERN and debuglog enabled using this testing script that did only two parallel requests:

```
#!/bin/bash

NPARALLEL=2
HERE=$PWD
cd /cvmfs/des.osgstorage.org/stash/kherner/load_testing
ORIGIN=http://fndca1.fnal.gov:8000
sudo cvmfs_talk -i des.osgstorage.org external host set "$ORIGIN;$ORIGIN;$ORIGIN;$ORIGIN"
mkdir -p $HERE/out
ACTIVEFILE=$HERE/out/$HOSTNAME.active
touch $ACTIVEFILE
trap "rm -f $ACTIVEFILE" 0
N=0
while [ $N -lt $NPARALLEL ]; do
    let N+=1
    find * -type f |shuf|while read F; do
	if [ ! -f $ACTIVEFILE ]; then
	    break
	fi
	date
	echo $F: $(cat $F|wc -c)
    done >$HERE/out/$HOSTNAME.$N 2>&1 &
done
wait
```

This is an edited subset of the debug log:
```
(cache) downloading /pnfs/fnal.gov/usr/des/persistent/stash/kherner/load_testing/skypca/y5e1/Y6A1_20170104t190116_u_c04_r4099p01_skypca-tmpl.fits    [10-09-2023 19:00:59 CEST]
(cache) start transaction on ./txn/fetch3AZ8rl has result 54    [10-09-2023 19:00:59 CEST]
(cache) miss: /pnfs/fnal.gov/usr/des/persistent/stash/kherner/load_testing/skypca/y5e1/Y6A1_20170104t190116_u_c04_r4099p01_skypca-tmpl.fits /pnfs/fnal.gov/usr/des/persistent/stash/kherner/load_testing/skypca/y5e1/Y6A1_20170104t190116_u_c04_r4099p01_skypca-tmpl.fits    [10-09-2023 19:00:59 CEST]
(download) escaped http://fndca1.fnal.gov:8000/pnfs/fnal.gov/usr/des/persistent/stash/kherner/load_testing/norm_dflatcor/y2e1/D_n20141020t1030_Y_c57_r1471p01_norm-dflatcor.fits to http://fndca1.fnal.gov:8000/pnfs/fnal.gov/usr/des/persistent/stash/kherner/load_testing/norm_dflatcor/y2e1/D_n20141020t1030_Y_c57_r1471p01_norm-dflatcor.fits    [10-09-2023 19:00:59 CEST]
(download) escaped http://fndca1.fnal.gov:8000/pnfs/fnal.gov/usr/des/persistent/stash/kherner/load_testing/skypca/y5e1/Y6A1_20170104t190116_u_c04_r4099p01_skypca-tmpl.fits to http://fndca1.fnal.gov:8000/pnfs/fnal.gov/usr/des/persistent/stash/kherner/load_testing/skypca/y5e1/Y6A1_20170104t190116_u_c04_r4099p01_skypca-tmpl.fits    [10-09-2023 19:00:59 CEST]
(curl) {header/out} GET /pnfs/fnal.gov/usr/des/persistent/stash/kherner/load_testing/norm_dflatcor/y2e1/D_n20141020t1030_Y_c57_r1471p01_norm-dflatcor.fits HTTP/1.1
Host: fndca1.fnal.gov:8000
Range: bytes=0-25165823
Accept: */*
Connection: Keep-Alive
User-Agent: cvmfs Fuse 2.11.0
cvmfs-info: /pnfs/fnal.gov/usr/des/persistent/stash/kherner/load_testing/norm_dflatcor/y2e1/D_n20141020t1030_Y_c57_r1471p01_norm-dflatcor.fits    [10-09-2023 19:00:59 CEST]
(curl) {header/out} GET /pnfs/fnal.gov/usr/des/persistent/stash/kherner/load_testing/skypca/y5e1/Y6A1_20170104t190116_u_c04_r4099p01_skypca-tmpl.fits HTTP/1.1
Host: fndca1.fnal.gov:8000
Range: bytes=0-25165823
Accept: */*
Connection: Keep-Alive
User-Agent: cvmfs Fuse 2.11.0
cvmfs-info: /pnfs/fnal.gov/usr/des/persistent/stash/kherner/load_testing/skypca/y5e1/Y6A1_20170104t190116_u_c04_r4099p01_skypca-tmpl.fits    [10-09-2023 19:00:59 CEST]
(curl) {header/in} HTTP/1.1 302 Found    [10-09-2023 19:00:59 CEST]
(download) http redirect: HTTP/1.1 302 Found
    [10-09-2023 19:00:59 CEST]
(curl) {header/in} Date: Mon, 09 Oct 2023 17:00:59 GMT    [10-09-2023 19:00:59 CEST]
(curl) {header/in} Server: dCache/8.2.23-FNAL    [10-09-2023 19:00:59 CEST]
(curl) {header/in} Strict-Transport-Security: max-age=31536000    [10-09-2023 19:00:59 CEST]
(curl) {header/in} Location: http://[2620:6a:0:4812:f0:0:69:198]:20413/usr/des/persistent/stash/kherner/load_testing/norm_dflatcor/y2e1/D_n20141020t1030_Y_c57_r1471p01_norm-dflatcor.fits?dcache-http-uuid=a5995bd0-dc1c-46bd-b462-24ae903b1764&dcache-http-ref=http%3A%2F%2Ffndca1.fnal.gov%3A8000    [10-09-2023 19:00:59 CEST]
(download) Location: http://[2620:6a:0:4812:f0:0:69:198]:20413/usr/des/persistent/stash/kherner/load_testing/norm_dflatcor/y2e1/D_n20141020t1030_Y_c57_r1471p01_norm-dflatcor.fits?dcache-http-uuid=a5995bd0-dc1c-46bd-b462-24ae903b1764&dcache-http-ref=http%3A%2F%2Ffndca1.fnal.gov%3A8000
    [10-09-2023 19:00:59 CEST]
(curl) {header/in} Content-Length: 0    [10-09-2023 19:00:59 CEST]
(curl) {header/in}     [10-09-2023 19:00:59 CEST]
(curl) {header/in} HTTP/1.1 302 Found    [10-09-2023 19:00:59 CEST]
(download) http redirect: HTTP/1.1 302 Found
    [10-09-2023 19:00:59 CEST]
(curl) {header/in} Date: Mon, 09 Oct 2023 17:00:59 GMT    [10-09-2023 19:00:59 CEST]
(curl) {header/in} Server: dCache/8.2.23-FNAL    [10-09-2023 19:00:59 CEST]
(curl) {header/in} Strict-Transport-Security: max-age=31536000    [10-09-2023 19:00:59 CEST]
(curl) {header/in} Location: http://[2620:6a:0:4812:f0:0:69:198]:20413/usr/des/persistent/stash/kherner/load_testing/skypca/y5e1/Y6A1_20170104t190116_u_c04_r4099p01_skypca-tmpl.fits?dcache-http-uuid=3bcc8d1a-4838-40d6-ab5c-0ed88425a666&dcache-http-ref=http%3A%2F%2Ffndca1.fnal.gov%3A8000    [10-09-2023 19:00:59 CEST]
(download) Location: http://[2620:6a:0:4812:f0:0:69:198]:20413/usr/des/persistent/stash/kherner/load_testing/skypca/y5e1/Y6A1_20170104t190116_u_c04_r4099p01_skypca-tmpl.fits?dcache-http-uuid=3bcc8d1a-4838-40d6-ab5c-0ed88425a666&dcache-http-ref=http%3A%2F%2Ffndca1.fnal.gov%3A8000
    [10-09-2023 19:00:59 CEST]
(curl) {header/in} Content-Length: 0    [10-09-2023 19:00:59 CEST]
(curl) {header/in}     [10-09-2023 19:00:59 CEST]
(curl) {header/out} GET /usr/des/persistent/stash/kherner/load_testing/norm_dflatcor/y2e1/D_n20141020t1030_Y_c57_r1471p01_norm-dflatcor.fits?dcache-http-uuid=a5995bd0-dc1c-46bd-b462-24ae903b1764&dcache-http-ref=http%3A%2F%2Ffndca1.fnal.gov%3A8000 HTTP/1.1
Host: [2620:6a:0:4812:f0:0:69:198]:20413
Range: bytes=0-25165823
Accept: */*
Connection: Keep-Alive
User-Agent: cvmfs Fuse 2.11.0
cvmfs-info: /pnfs/fnal.gov/usr/des/persistent/stash/kherner/load_testing/norm_dflatcor/y2e1/D_n20141020t1030_Y_c57_r1471p01_norm-dflatcor.fits    [10-09-2023 19:01:00 CEST]
(curl) {header/out} GET /usr/des/persistent/stash/kherner/load_testing/skypca/y5e1/Y6A1_20170104t190116_u_c04_r4099p01_skypca-tmpl.fits?dcache-http-uuid=3bcc8d1a-4838-40d6-ab5c-0ed88425a666&dcache-http-ref=http%3A%2F%2Ffndca1.fnal.gov%3A8000 HTTP/1.1
Host: [2620:6a:0:4812:f0:0:69:198]:20413
Range: bytes=0-25165823
Accept: */*
Connection: Keep-Alive
User-Agent: cvmfs Fuse 2.11.0
cvmfs-info: /pnfs/fnal.gov/usr/des/persistent/stash/kherner/load_testing/skypca/y5e1/Y6A1_20170104t190116_u_c04_r4099p01_skypca-tmpl.fits    [10-09-2023 19:01:00 CEST]
(curl) {header/in} HTTP/1.1 206 Partial Content    [10-09-2023 19:01:00 CEST]
(curl) {header/in} Accept-Ranges: bytes    [10-09-2023 19:01:00 CEST]
(curl) {header/in} Content-Length: 25165824    [10-09-2023 19:01:00 CEST]
(curl) {header/in} Content-Range: bytes 0-25165823/134222400    [10-09-2023 19:01:00 CEST]
(curl) {header/in} Server: dCache/8.2.23-FNAL    [10-09-2023 19:01:00 CEST]
(curl) {header/in}     [10-09-2023 19:01:00 CEST]
(curl) {header/in} HTTP/1.1 206 Partial Content    [10-09-2023 19:01:00 CEST]
(curl) {header/in} Accept-Ranges: bytes    [10-09-2023 19:01:00 CEST]
(curl) {header/in} Content-Length: 25165824    [10-09-2023 19:01:00 CEST]
(curl) {header/in} Content-Range: bytes 0-25165823/67124160    [10-09-2023 19:01:00 CEST]
(curl) {header/in} Server: dCache/8.2.23-FNAL    [10-09-2023 19:01:00 CEST]
(curl) {header/in}     [10-09-2023 19:01:00 CEST]
(curl) {data/in} <snip>    [10-09-2023 19:01:00 CEST]
...
(download) Verify downloaded url /pnfs/fnal.gov/usr/des/persistent/stash/kherner/load_testing/norm_dflatcor/y2e1/D_n20141020t1030_Y_c57_r1471p01_norm-dflatcor.fits, proxy DIRECT (curl error 0)    [10-09-2023 19:01:31 CEST]
(cache) finished downloading of /pnfs/fnal.gov/usr/des/persistent/stash/kherner/load_testing/norm_dflatcor/y2e1/D_n20141020t1030_Y_c57_r1471p01_norm-dflatcor.fits    [10-09-2023 19:01:31 CEST]
(cache) commit ./24/5f7c9ba23aa812ce77dc3fcf13d1be21d83cf7 ./txn/fetch5EWY79    [10-09-2023 19:01:31 CEST]
(quota) insert into lru 245f7c9ba23aa812ce77dc3fcf13d1be21d83cf7, path Part of /pnfs/fnal.gov/usr/des/persistent/stash/kherner/load_testing/norm_dflatcor/y2e1/D_n20141020t1030_Y_c57_r1471p01_norm-dflatcor.fits, method 1    [10-09-2023 19:01:31 CEST]
(cvmfs) reading from chunk fd 61    [10-09-2023 19:01:31 CEST]
(cvmfs) released chunk file descriptor 61    [10-09-2023 19:01:31 CEST]
(cvmfs) pushed 131072 bytes to user    [10-09-2023 19:01:31 CEST]
...
(cache) downloading /pnfs/fnal.gov/usr/des/persistent/stash/kherner/load_testing/norm_dflatcor/y2e1/D_n20141020t1030_Y_c57_r1471p01_norm-dflatcor.fits    [10-09-2023 19:01:31 CEST]
(cvmfs) cvmfs_read inode: 173243 reading 131072 bytes from offset 25296896 fd -2    [10-09-2023 19:01:31 CEST]
(cache) start transaction on ./txn/fetchpMBSiP has result 49    [10-09-2023 19:01:31 CEST]
(cvmfs) cvmfs_read inode: 173243 reading 131072 bytes from offset 25427968 fd -2    [10-09-2023 19:01:31 CEST]
(cache) miss: /pnfs/fnal.gov/usr/des/persistent/stash/kherner/load_testing/norm_dflatcor/y2e1/D_n20141020t1030_Y_c57_r1471p01_norm-dflatcor.fits /pnfs/fnal.gov/usr/des/persistent/stash/kherner/load_testing/norm_dflatcor/y2e1/D_n20141020t1030_Y_c57_r1471p01_norm-dflatcor.fits    [10-09-2023 19:01:31 CEST]
(download) escaped http://fndca1.fnal.gov:8000/pnfs/fnal.gov/usr/des/persistent/stash/kherner/load_testing/norm_dflatcor/y2e1/D_n20141020t1030_Y_c57_r1471p01_norm-dflatcor.fits to http://fndca1.fnal.gov:8000/pnfs/fnal.gov/usr/des/persistent/stash/kherner/load_testing/norm_dflatcor/y2e1/D_n20141020t1030_Y_c57_r1471p01_norm-dflatcor.fits    [10-09-2023 19:01:31 CEST]
(curl) {header/out} GET /pnfs/fnal.gov/usr/des/persistent/stash/kherner/load_testing/norm_dflatcor/y2e1/D_n20141020t1030_Y_c57_r1471p01_norm-dflatcor.fits HTTP/1.1
Host: fndca1.fnal.gov:8000
Range: bytes=25165824-50331647
Accept: */*
Connection: Keep-Alive
User-Agent: cvmfs Fuse 2.11.0
cvmfs-info: /pnfs/fnal.gov/usr/des/persistent/stash/kherner/load_testing/norm_dflatcor/y2e1/D_n20141020t1030_Y_c57_r1471p01_norm-dflatcor.fits    [10-09-2023 19:01:31 CEST]
(curl) {data/in} <snip>    [10-09-2023 19:01:31 CEST]
...
(curl) {header/in} HTTP/1.1 302 Found    [10-09-2023 19:01:31 CEST]
(download) http redirect: HTTP/1.1 302 Found
    [10-09-2023 19:01:31 CEST]
(curl) {header/in} Date: Mon, 09 Oct 2023 17:01:31 GMT    [10-09-2023 19:01:31 CEST]
(curl) {header/in} Server: dCache/8.2.23-FNAL    [10-09-2023 19:01:31 CEST]
(curl) {header/in} Strict-Transport-Security: max-age=31536000    [10-09-2023 19:01:31 CEST]
(curl) {header/in} Location: http://[2620:6a:0:4812:f0:0:69:198]:20413/usr/des/persistent/stash/kherner/load_testing/norm_dflatcor/y2e1/D_n20141020t1030_Y_c57_r1471p01_norm-dflatcor.fits?dcache-http-uuid=aac8b36e-d5ae-4ec8-a6c9-a3a0278be979&dcache-http-ref=http%3A%2F%2Ffndca1.fnal.gov%3A8000    [10-09-2023 19:01:31 CEST]
(download) Location: http://[2620:6a:0:4812:f0:0:69:198]:20413/usr/des/persistent/stash/kherner/load_testing/norm_dflatcor/y2e1/D_n20141020t1030_Y_c57_r1471p01_norm-dflatcor.fits?dcache-http-uuid=aac8b36e-d5ae-4ec8-a6c9-a3a0278be979&dcache-http-ref=http%3A%2F%2Ffndca1.fnal.gov%3A8000
    [10-09-2023 19:01:31 CEST]
(curl) {header/in} Content-Length: 0    [10-09-2023 19:01:31 CEST]
(curl) {header/in}     [10-09-2023 19:01:31 CEST]
(curl) {header/out} GET /usr/des/persistent/stash/kherner/load_testing/norm_dflatcor/y2e1/D_n20141020t1030_Y_c57_r1471p01_norm-dflatcor.fits?dcache-http-uuid=aac8b36e-d5ae-4ec8-a6c9-a3a0278be979&dcache-http-ref=http%3A%2F%2Ffndca1.fnal.gov%3A8000 HTTP/1.1
Host: [2620:6a:0:4812:f0:0:69:198]:20413
Range: bytes=25165824-50331647
Accept: */*
Connection: Keep-Alive
User-Agent: cvmfs Fuse 2.11.0
cvmfs-info: /pnfs/fnal.gov/usr/des/persistent/stash/kherner/load_testing/norm_dflatcor/y2e1/D_n20141020t1030_Y_c57_r1471p01_norm-dflatcor.fits    [10-09-2023 19:01:31 CEST]
...
(download) Verify downloaded url /pnfs/fnal.gov/usr/des/persistent/stash/kherner/load_testing/skypca/y5e1/Y6A1_20170104t190116_u_c04_r4099p01_skypca-tmpl.fits, proxy DIRECT (curl error 0)    [10-09-2023 19:01:32 CEST]
(cache) finished downloading of /pnfs/fnal.gov/usr/des/persistent/stash/kherner/load_testing/skypca/y5e1/Y6A1_20170104t190116_u_c04_r4099p01_skypca-tmpl.fits    [10-09-2023 19:01:32 CEST]
(cache) commit ./19/0889782281f9441de07d11b9ed2db1554f9cee ./txn/fetch3AZ8rl    [10-09-2023 19:01:32 CEST]
(quota) insert into lru 190889782281f9441de07d11b9ed2db1554f9cee, path Part of /pnfs/fnal.gov/usr/des/persistent/stash/kherner/load_testing/skypca/y5e1/Y6A1_20170104t190116_u_c04_r4099p01_skypca-tmpl.fits, method 1    [10-09-2023 19:01:32 CEST]
(cvmfs) reading from chunk fd 61    [10-09-2023 19:01:32 CEST]
(cvmfs) released chunk file descriptor 61    [10-09-2023 19:01:32 CEST]
(cvmfs) pushed 131072 bytes to user    [10-09-2023 19:01:32 CEST]
...
(curl) {header/out} GET /usr/des/persistent/stash/kherner/load_testing/norm_dflatcor/y2e1/D_n20141020t1030_Y_c57_r1471p01_norm-dflatcor.fits?dcache-http-uuid=aac8b36e-d5ae-4ec8-a6c9-a3a0278be979&dcache-http-ref=http%3A%2F%2Ffndca1.fnal.gov%3A8000 HTTP/1.1
Host: [2620:6a:0:4812:f0:0:69:198]:20413
Range: bytes=25165824-50331647
Accept: */*
Connection: Keep-Alive
User-Agent: cvmfs Fuse 2.11.0
cvmfs-info: /pnfs/fnal.gov/usr/des/persistent/stash/kherner/load_testing/norm_dflatcor/y2e1/D_n20141020t1030_Y_c57_r1471p01_norm-dflatcor.fits    [10-09-2023 19:01:32 CEST]
(cvmfs) pushed 131072 bytes to user    [10-09-2023 19:01:32 CEST]
(cvmfs) cvmfs_read inode: 179029 reading 131072 bytes from offset 5373952 fd -3    [10-09-2023 19:01:32 CEST]
...
(cache) downloading /pnfs/fnal.gov/usr/des/persistent/stash/kherner/load_testing/skypca/y5e1/Y6A1_20170104t190116_u_c04_r4099p01_skypca-tmpl.fits    [10-09-2023 19:01:32 CEST]
(cvmfs) cvmfs_read inode: 179029 reading 131072 bytes from offset 25296896 fd -3    [10-09-2023 19:01:32 CEST]
(cache) start transaction on ./txn/fetchPNBE6j has result 62    [10-09-2023 19:01:32 CEST]
(cache) miss: /pnfs/fnal.gov/usr/des/persistent/stash/kherner/load_testing/skypca/y5e1/Y6A1_20170104t190116_u_c04_r4099p01_skypca-tmpl.fits /pnfs/fnal.gov/usr/des/persistent/stash/kherner/load_testing/skypca/y5e1/Y6A1_20170104t190116_u_c04_r4099p01_skypca-tmpl.fits    [10-09-2023 19:01:32 CEST]
(cvmfs) cvmfs_read inode: 179029 reading 131072 bytes from offset 25427968 fd -3    [10-09-2023 19:01:32 CEST]
(download) escaped http://fndca1.fnal.gov:8000/pnfs/fnal.gov/usr/des/persistent/stash/kherner/load_testing/skypca/y5e1/Y6A1_20170104t190116_u_c04_r4099p01_skypca-tmpl.fits to http://fndca1.fnal.gov:8000/pnfs/fnal.gov/usr/des/persistent/stash/kherner/load_testing/skypca/y5e1/Y6A1_20170104t190116_u_c04_r4099p01_skypca-tmpl.fits    [10-09-2023 19:01:32 CEST]
(curl) {header/out} GET /pnfs/fnal.gov/usr/des/persistent/stash/kherner/load_testing/skypca/y5e1/Y6A1_20170104t190116_u_c04_r4099p01_skypca-tmpl.fits HTTP/1.1
Host: fndca1.fnal.gov:8000
Range: bytes=25165824-50331647
Accept: */*
Connection: Keep-Alive
User-Agent: cvmfs Fuse 2.11.0
cvmfs-info: /pnfs/fnal.gov/usr/des/persistent/stash/kherner/load_testing/skypca/y5e1/Y6A1_20170104t190116_u_c04_r4099p01_skypca-tmpl.fits    [10-09-2023 19:01:32 CEST]
(curl) {header/in} HTTP/1.1 307 Request is no longer valid    [10-09-2023 19:01:32 CEST]
(download) http redirect: HTTP/1.1 307 Request is no longer valid
    [10-09-2023 19:01:32 CEST]
(curl) {header/in} Content-Length: 0    [10-09-2023 19:01:32 CEST]
(curl) {header/in} Location: http://fndca1.fnal.gov:8000/usr/des/persistent/stash/kherner/load_testing/norm_dflatcor/y2e1/D_n20141020t1030_Y_c57_r1471p01_norm-dflatcor.fits    [10-09-2023 19:01:32 CEST]
(download) Location: http://fndca1.fnal.gov:8000/usr/des/persistent/stash/kherner/load_testing/norm_dflatcor/y2e1/D_n20141020t1030_Y_c57_r1471p01_norm-dflatcor.fits
    [10-09-2023 19:01:32 CEST]
(curl) {header/in} Server: dCache/8.2.23-FNAL    [10-09-2023 19:01:32 CEST]
(curl) {header/in}     [10-09-2023 19:01:32 CEST]
(curl) {header/out} GET /usr/des/persistent/stash/kherner/load_testing/norm_dflatcor/y2e1/D_n20141020t1030_Y_c57_r1471p01_norm-dflatcor.fits HTTP/1.1
Host: fndca1.fnal.gov:8000
Range: bytes=25165824-50331647
Accept: */*
Connection: Keep-Alive
User-Agent: cvmfs Fuse 2.11.0
cvmfs-info: /pnfs/fnal.gov/usr/des/persistent/stash/kherner/load_testing/norm_dflatcor/y2e1/D_n20141020t1030_Y_c57_r1471p01_norm-dflatcor.fits    [10-09-2023 19:01:32 CEST]
(curl) {header/in} HTTP/1.1 302 Found    [10-09-2023 19:01:32 CEST]
(download) http redirect: HTTP/1.1 302 Found
    [10-09-2023 19:01:32 CEST]
(curl) {header/in} Date: Mon, 09 Oct 2023 17:01:32 GMT    [10-09-2023 19:01:32 CEST]
(curl) {header/in} Server: dCache/8.2.23-FNAL    [10-09-2023 19:01:32 CEST]
(curl) {header/in} Strict-Transport-Security: max-age=31536000    [10-09-2023 19:01:32 CEST]
(curl) {header/in} Location: http://[2620:6a:0:4812:f0:0:69:198]:20413/usr/des/persistent/stash/kherner/load_testing/skypca/y5e1/Y6A1_20170104t190116_u_c04_r4099p01_skypca-tmpl.fits?dcache-http-uuid=1219de7b-37e5-4e46-8cc2-b7d44f3144e9&dcache-http-ref=http%3A%2F%2Ffndca1.fnal.gov%3A8000    [10-09-2023 19:01:32 CEST]
(download) Location: http://[2620:6a:0:4812:f0:0:69:198]:20413/usr/des/persistent/stash/kherner/load_testing/skypca/y5e1/Y6A1_20170104t190116_u_c04_r4099p01_skypca-tmpl.fits?dcache-http-uuid=1219de7b-37e5-4e46-8cc2-b7d44f3144e9&dcache-http-ref=http%3A%2F%2Ffndca1.fnal.gov%3A8000
    [10-09-2023 19:01:32 CEST]
(curl) {header/in} Content-Length: 0    [10-09-2023 19:01:32 CEST]
(curl) {header/in}     [10-09-2023 19:01:32 CEST]
(curl) {header/out} GET /usr/des/persistent/stash/kherner/load_testing/skypca/y5e1/Y6A1_20170104t190116_u_c04_r4099p01_skypca-tmpl.fits?dcache-http-uuid=1219de7b-37e5-4e46-8cc2-b7d44f3144e9&dcache-http-ref=http%3A%2F%2Ffndca1.fnal.gov%3A8000 HTTP/1.1
Host: [2620:6a:0:4812:f0:0:69:198]:20413
Range: bytes=25165824-50331647
Accept: */*
Connection: Keep-Alive
User-Agent: cvmfs Fuse 2.11.0
cvmfs-info: /pnfs/fnal.gov/usr/des/persistent/stash/kherner/load_testing/skypca/y5e1/Y6A1_20170104t190116_u_c04_r4099p01_skypca-tmpl.fits    [10-09-2023 19:01:32 CEST]
(curl) {header/in} HTTP/1.1 401 Unauthorized    [10-09-2023 19:01:32 CEST]
(download) http status error code: HTTP/1.1 401 Unauthorized
 [401]    [10-09-2023 19:01:32 CEST]
(curl) {info} Failed writing header    [10-09-2023 19:01:32 CEST]
(download) Verify downloaded url /pnfs/fnal.gov/usr/des/persistent/stash/kherner/load_testing/norm_dflatcor/y2e1/D_n20141020t1030_Y_c57_r1471p01_norm-dflatcor.fits, proxy DIRECT (curl error 23)    [10-09-2023 19:01:32 CEST]
(download) Trying again on same curl handle, same url: 0, error code 9 no-cache 0    [10-09-2023 19:01:32 CEST]
(download) switching host from http://fndca1.fnal.gov:8000 to http://fndca1.fnal.gov:8000 (host returned HTTP error)    [10-09-2023 19:01:32 CEST]
```
It's clear that in this case that two different files `skypca/y5e1/Y6A1_20170104t190116_u_c04_r4099p01_skypca-tmpl.fits` and `norm_dflatcor/y2e1/D_n20141020t1030_Y_c57_r1471p01_norm-dflatcor.fits` are being dealt with in an interleaved fashion.  That by itself is not a problem, it just causes debugging to be difficult because the two different requests and responses show up mixed up in the output.  However I do believe that a clear problem is shown in the fact that this line appears twice:
```
(curl) {header/out} GET /usr/des/persistent/stash/kherner/load_testing/norm_dflatcor/y2e1/D_n20141020t1030_Y_c57_r1471p01_norm-dflatcor.fits?dcache-http-uuid=aac8b36e-d5ae-4ec8-a6c9-a3a0278be979&dcache-http-ref=http%3A%2F%2Ffndca1.fnal.gov%3A8000 HTTP/1.1
```
Both of them are followed by `Range: bytes=25165824-50331647` so it really does seem to be an attempt to follow the same redirect twice.  dCache does not allow that so the second one gets `307 Request is no longer valid` response.  I only see one case where the redirect (that is `302 Not Found`) has a `Location` header that says to read this URL.  It is in the log immediately before that first GET of the URL.  In between the two GETs of the URL it says `finished downloading` of the other file, which I expect has something to do with causing the problem.

I haven't verified it but suspect that the xrootd server http interface does not use the 302 Not Found response to redirect,  so this problem probably only occurs when `CVMFS_FOLLOW_REDIRECTS=yes` is needed as it is for dCache.

**Comment by jblomer** [comment]
Interesting! When I follow the log, I come to the same conclusion. The redirects are handled directly by curl though. We are investigating.

**Comment by DrDaveD** [comment]
Any progress on this?  It is causing secondary problems at Fermilab, because we are apparently overloading the stashcache servers.

**Comment by jblomer** [comment]
Unfortunately not yet. It's the next thing we are addressing.

**Comment by DrDaveD** [comment]
How about since then?  This is still causing a lot of problems for us.

**Comment by DrDaveD** [comment]
Laura looked into this and sent me wireshark traces that implicated the server side.  Libcurl was re-requesting the data because an empty response was returned from the first request.

I brought that to the attention of our dCache expert and he found exceptions in his logs like this:
```
30 Jan 2024 11:03:36 (p-des-pubstor2232-5) [door:WebDAV-fndca4b-2@webdavDomain1:AAYQLL4e5Ig WebDAV-fndca4b-2 PoolDeliverFile 000041C8AD9299CE4E389159A1DCD949D1C1] Transfer failed: java.util.NoSuchElementException: io.netty.handler.timeout.IdleStateHandler
```
He upgraded the dCache software in the data pool nodes being used (since they were dedicated for testing) to version 8.2.27 from 8.2.23, and the problem went away.  The rest of the cluster should get upgraded later this month.  He said the exception was triggered under other conditions and fixed for that reason, but cvmfs was also triggering the same condition.  

Bottom line, it's not a cvmfs problem after all.  Closing the issue.  Thanks a lot for your investigation.
